Upgrade to Apple Silicon on CI, simplify Apple Silicon installation instructions

nilsvu opened the pull request:
## Proposed changes

Upgrade the macOS CI test to run on Apple Silicon. Also simplify Apple Silicon installation instructions a bit.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Remove libsharp & brigand from installation instructions, container

These are bundled now.

nilsdeppe reviewed (dismissed):
Thank you!

knelli2 commented:
MacOS spack installation of pybind11 isn't happy

knelli2 commented on .github/workflows/Tests.yaml:
```diff
@@ -983,6 +976,7 @@ ${{ matrix.build_type }}-pch-${{ matrix.use_pch || 'ON' }}"
             -D OVERRIDE_ARCH=x86-64 \
             -D BUILD_SHARED_LIBS=ON \
             -D BUILD_PYTHON_BINDINGS=ON \
+            -D BOOTSTRAP_PY_DEPS=ON \
```
The `SetupPybind11.cmake` isn't finding the boostrapped pybind11. Maybe we need another hint in cmake?

nilsvu committed:
Upgrade macOS CI to Apple Silicon

nilsvu commented:
@geoffrey4444 FYI. Still needs #5956 to be merged first.

geoffrey4444 commented:
I'm excited about adding Apple Silicon to the CI...thanks for doing this!

I understand there are complexities to #5956 , so if that looks like it won't get merged for a while, maybe consider just implementing the slightly longer instructions in a separate PR, so we can start automatically checking Apple Silicon? We can always improve things later, but we have enough Apple Silicon users that it would be nice to get the CI up and running. But if $5956 is going to get merged soon-ish, maybe a separate PR isn't worth it.

For those who prefer manual installation (I'm probably one of them, at least for now), maybe instead of deleting the longer instructions, move them down to a "manually install dependencies" section?

nilsvu committed:
Update Apple Silicon installation instructions

nilsvu commented:
@geoffrey4444 ok I removed the dependence on #5956 for now. Once it's merged we can simplify the instructions further, but I agree that it would be good to keep an (optional) "install manually" step.

geoffrey4444 commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -64,31 +57,31 @@ pip install jupyterlab
-brew install boost gsl cmake doxygen catch2 openblas
-brew install ccache autoconf automake jemalloc hdf5 pybind11 yaml-cpp
+brew install gcc autoconf automake ccache cmake
+brew install boost catch2 doxygen gsl hdf5 jemalloc openblas yaml-cpp
```
Why no pybind11 here? Is this taken care of by step 2 above (i.e., the requirements or dev_requirements)?

nilsvu replied:
yes, that's one of the earlier commits in this PR.

geoffrey4444 commented on .github/workflows/Tests.yaml:
```diff
@@ -858,17 +858,15 @@ ${{ matrix.build_type }}-pch-${{ matrix.use_pch || 'ON' }}"
   # Build all test executables and run unit tests on macOS
   unit_tests_macos:
```
Can you please make 2 macOS tests: i) Apple Silicon, and ii) Intel?

nilsvu replied:
Oh I don't think we need Intel anymore, do we?

nilsdeppe replied:
People still use Intel Macs. They aren't that old yet so I think we should continue to test them

geoffrey4444 reviewed (changes requested):
Thanks again for doing this, @nilsvu ! I reviewed the Apple-specific commits (the last two). Is it possible to just spin them off into a separate PR, to get the Apple Silicon docker container merged ASAP? The other changes don't seem to affect the Apple Silicon instructions or adding Apple Silicon to the docker container.

nilsvu commented:
@nilsdeppe @geoffrey4444 @knelli2 I rebased and updated this. Please, could whoever of you is interested in this take another look so we can get this in? It would be helpful to have an ARM macOS build on CI.

knelli2 commented on docs/Installation/InstallationOnAppleSilicon.md:
```diff
@@ -158,21 +161,26 @@ cmake \
 -D BUILD_SHARED_LIBS=ON \
 -D MEMORY_ALLOCATOR=SYSTEM \
 -D CHARM_ROOT=${SPECTRE_DEPS_ROOT}/charm/multicore-darwin-arm8 \
+-D SPECTRE_FETCH_MISSING_DEPS=ON \
```
I guess it's up to @geoffrey4444 if he wants the docs default to install extra dependencies in people's build directories

geoffrey4444 replied:
I think fetching missing dependencies is fine

knelli2 reviewed (commented):
Changes look ok to me, but I don't have a mac to test.

nilsvu committed:
Loosen tolerances

This test failed on Intel macOS with OpenBLAS.

geoffrey4444 commented:
LGTM, but I notice it has conflicts that must be resolved first

nilsvu commented:
Rebased!

knelli2 commented:
Tolerances on the partial derivative test still seem to not be adequate.

nilsdeppe commented:
No need for my approval on this. I was pinged in the past, but have nothing to add.

nilsvu commented:
Tolerances seem to work now @knelli2
